gnucash/import-export/import-main-matcher.c

```diff
@@ -1352,26 +1354,116 @@ void gnc_gen_trans_list_add_trans_with_ref_id (GNCImportMainMatcher *gui, Transa
+{
+    GList* list_element = NULL;
+    GtkTreeIter iter;
+    Account *import_trans_account;
+    time64 import_trans_time, min_time=G_MAXINT64, max_time=0;
+    GList* all_accounts = NULL;
+
+    // Go through all imported transactions, gather the list of accounts, and min/max date range.
+    gboolean valid = gtk_tree_model_get_iter_first (model, &iter);
+    while(valid)
+    {
+        GNCImportTransInfo* transaction_info;
+        gtk_tree_model_get (model, &iter,
+                            DOWNLOADED_COL_DATA, &transaction_info,
+                            -1);
+        import_trans_account = xaccSplitGetAccount (gnc_import_TransInfo_get_fsplit (transaction_info));
+        all_accounts = g_list_prepend (all_accounts, import_trans_account);
+        import_trans_time = xaccTransGetDate (gnc_import_TransInfo_get_trans (transaction_info));
+        min_time = MIN (min_time, import_trans_time);
+        max_time = MAX (max_time, import_trans_time);
+
+        valid = gtk_tree_model_iter_next (model, &iter);
+    }
 
-        gnc_import_TransInfo_init_matches (transaction_info,
-                                           gui->user_settings);
+    // Make a query to find splits with the right accounts and dates.
+    qof_query_set_book (query, gnc_get_current_book());
+    xaccQueryAddAccountMatch (query, all_accounts,
+                              QOF_GUID_MATCH_ANY, QOF_QUERY_AND);
+    xaccQueryAddDateMatchTT (query,
+                             TRUE, min_time - match_date_hardlimit * 86400,
+                             TRUE, max_time + match_date_hardlimit * 86400,
+                             QOF_QUERY_AND);
+    list_element = qof_query_run (query);
+    g_list_free (all_accounts);
+    return list_element;
+}
+
+static void
+gnc_gen_trans_list_finalize (GNCImportMainMatcher *gui)
+{
+    GtkTreeModel* model = gtk_tree_view_get_model (gui->view);
+    GtkListStore* store = GTK_LIST_STORE (model);
+    GtkTreeIter iter;
+    GNCImportMatchInfo *selected_match;
+    gboolean match_selected_manually;
+    Account *importaccount;
+    GList* list_element = NULL;
+    GList* iter_2 = NULL;
+    Query *query;
+    gint match_date_hardlimit = gnc_import_Settings_get_match_date_hardlimit (gui->user_settings);
+    gint display_threshold = gnc_import_Settings_get_display_threshold (gui->user_settings);
+    double fuzzy_amount = gnc_import_Settings_get_fuzzy_amount (gui->user_settings);
+    gboolean valid;
+
+    query = qof_query_create_for (GNC_ID_SPLIT);
+    // Gather the list of splits that could match one of the imported transactions
+    list_element = create_list_of_potential_matches (model, query, match_date_hardlimit);
+
+    // For each imported transaction, go through list_element, select the ones that have the
+    // right account, and evaluate how good a match they are.
+    valid = gtk_tree_model_get_iter_first (model, &iter);
+    while(valid)
+    {
```

Comment:
A `for` loop (see below) is better, but `gtk_tree_model_foreach()` would be better still.

Reply:
gtk_tree_model_foreach() requires defining a function, IMHO the added function obscures the code a bit, for a gain that's pretty minimal. But if you insist, I'll do it, no problem.

Reply:
We use the `foo_for_each` idiom quite a bit in GnuCash, and the STL equivalent is what's taught as The Right Thing To Do™ in the C++ universe nowadays. In general decomposing to functions and algorithms has been considered a fundamental [Good Thing™](https://martinfowler.com/books/refactoring.html) for at least 20 years.

Unlike the STL case there's probably no performance advantage over using a naked loop so you can take it as a suggestion instead of requirement. If you decide not to then do replace the `while`s with `for`s.

Reply:
I also prefer `for-each` much more than `do..while` or `while`.
This branch will also speed up all imports except QIF, right?

Reply:
> Unlike the STL case there's probably no performance advantage over using a naked loop
Not sure I understand you @jralls - are you saying in C++ `std::for_each` might introduce a performance penalty in comparison to naked loop? It will be a bit of surprise to see that in optimized builds.

Reply:
Here's my beef with gtk_tree_model_foreach():  I need to pass a bunch of auxiliary data to the function that will be called, but GTK only provides a single pointer. So I'd have to pack everything into a structure, and pass the pointer to that structure. On the function side, I'd need to unpack the structure into the data I need.
That, to me, is ugly as hell, and bloats the code. Not sure about the stl version but some foreach constructs allow passing a variable number of arguments so you don't have that issue. Not so here.

Reply:
>  @jralls - are you saying in C++ std::for_each might introduce a performance penalty in comparison to naked loop? It will be a bit of surprise to see that in optimized builds.

Read what you quoted more carefully: ***Unlike** the STL case*...

The STL algorithms can access private data and functions in STL containers that naked loops cannot, STL implementations are closely tied to their associated compilers and so can make use of compiler features both public and private that portable code cannot, and the STL implementors have time and motivation to focus on wringing the best possible performance out of the standard algorithms.

GLib's foreach functions have none of those advantages and some of them suffer from what Dr. Knuth calls "premature pessimization". Unfortunately profiling has shown that just about everything associated with GtkTreeModel iteration performs pretty poorly but there's no way around that if one wants to use a GtkTreeView.

Reply:
@bhardwajs 
  >  @jralls - are you saying in C++ std::for_each might introduce a performance penalty in comparison to naked loop? It will be a bit of surprise to see that in optimized builds.

Read what you quoted more carefully: Unlike the STL case...

The STL algorithms can access private data and functions in STL containers that naked loops cannot, STL implementations are closely tied to their associated compilers and so can make use of compiler features both public and private that portable code cannot, and the STL implementors have time and motivation to focus on wringing the best possible performance out of the standard algorithms.

GLib's foreach functions have none of those advantages and some of them suffer from what Dr. Knuth calls "premature pessimization". Unfortunately profiling has shown that just about everything associated with GtkTreeModel iteration performs pretty poorly but there's no way around that if one wants to use a GtkTreeView.

Reply:
@jeanlaroche Your antipathy to structs is misplaced; varargs are a much heavier-weight way of passing multiple parameters.

You don't "unpack" structs. The only extra effort to using a parameter in a struct is typing `foo->bar` instead of `bar`.

No struct needed in this case, though: the function would need only `list_item` or the hashtable I proposed.

Reply:
> No struct needed in this case, though: the function would need only `list_item` or the hashtable I proposed.

I need model, list_element, display_threshold, fuzzy_amount, and gui

Reply:
Not `model`, it's passed to the `GtkTreeModelForeachFunc` by `gtk_tree_model_for_each` along with `path`, which you don't need and `iter`, which you do. I'll concede the other 3, I lost track of where the loop stopped.